crates/pallet-domains/src/lib.rs

```diff
@@ -656,3 +765,36 @@ where
         }
     }
 }
+
+/// Calculates the new tx range based on the bundles produced during the interval.
+pub fn calculate_tx_range(
+    cur_tx_range: U256,
+    actual_bundle_count: u64,
+    expected_bundle_count: u64,
+) -> U256 {
+    if actual_bundle_count == 0 || expected_bundle_count == 0 {
+        return cur_tx_range;
+    }
+
+    let actual = U256::from(actual_bundle_count);
+    let expected = U256::from(expected_bundle_count);
+    let new_tx_range = actual.saturating_mul(&cur_tx_range);
+    let new_tx_range = if let Some(div) = new_tx_range.checked_div(&expected) {
```

Comment:
In order to make `calculate_tx_range` more readable (stay closely to the formula in the spec), we could add a new method `clamp` in U256, and make use of the `let ... else ...` syntax sugar.

```rust
    let Some(new_tx_range) = U256::from(actual_bundle_count)
        .saturating_mul(&current_tx_range)
        .checked_div(&U256::from(expected_bundle_count)) else {
        return cur_tx_range;
    };

    let upper_bound = cur_tx_range.saturating_mul(&U256::from(4_u64));
    let Some(lower_bound) = cur_tx_range.checked_div(&U256::from(4_u64)) else {
        return cur_tx_range;
    };

    new_tx_range.clamp(lower_bound, upper_bound)
```

Reply:
It turns out clamp is auto implemented if Ord trait is derived: https://doc.rust-lang.org/std/cmp/trait.Ord.html#method.clamp, so we don't really have to change our U256.